portal-kernel/test/unit/com/liferay/portal/kernel/util/ClassUtilTest.java

```diff
@@ -54,6 +88,88 @@ public void testGetClassesFromAnnotationsWithArrayParameter()
 			"AnnotationClass.Annotation", "AnnotationClass", "A", "B", "C");
 	}
 
+	@Test
+	public void testGetPathFromURLUnix() throws Exception {
+		testGetPathFromURL(
+			"jar:file:/opt/liferay/tomcat/lib/servlet-api.jar" +
+				"!/javax/servlet/Servlet.class",
+			"/opt/liferay/tomcat/lib/servlet-api.jar" +
+				"!/javax/servlet/Servlet.class");
+		testGetPathFromURL(
+			"jar:file:/opt/with%20space/tomcat/lib/servlet-api.jar" +
+				"!/javax/servlet/Servlet.class",
+			"/opt/with space/tomcat/lib/servlet-api.jar" +
+				"!/javax/servlet/Servlet.class");
+		testGetPathFromURL(
+			"file:/opt/liferay/tomcat/classes/javax/servlet/Servlet.class",
+			"/opt/liferay/tomcat/classes/javax/servlet/Servlet.class");
+	}
+
+	@Test
+	public void testGetPathFromURLWeblogic() throws Exception {
+		testGetPathFromURL(
+			"zip:/opt/liferay/tomcat/lib/servlet-api.jar" +
+				"!/javax/servlet/Servlet.class",
+			"/opt/liferay/tomcat/lib/servlet-api.jar" +
+				"!/javax/servlet/Servlet.class");
+		testGetPathFromURL(
+			"zip:/opt/with%20space/tomcat/lib/servlet-api.jar" +
+				"!/javax/servlet/Servlet.class",
+			"/opt/with space/tomcat/lib/servlet-api.jar" +
+				"!/javax/servlet/Servlet.class");
+		testGetPathFromURL(
+			"zip:C:/Liferay/tomcat/lib/servlet-api.jar" +
+				"!/javax/servlet/Servlet.class",
+			"/C:/Liferay/tomcat/lib/servlet-api.jar" +
+				"!/javax/servlet/Servlet.class");
+		testGetPathFromURL(
+			"zip:C:/With%20Space/tomcat/lib/servlet-api.jar" +
+				"!/javax/servlet/Servlet.class",
+			"/C:/With Space/tomcat/lib/servlet-api.jar" +
+				"!/javax/servlet/Servlet.class");
+	}
+
+	@Test
+	public void testGetPathFromURLWildfly() throws Exception {
+		testGetPathFromURL(
+			"vfs:/opt/liferay/tomcat/lib/servlet-api.jar/javax/servlet" +
+				"/Servlet.class",
+			"/opt/liferay/tomcat/lib/servlet-api.jar/javax/servlet" +
+				"/Servlet.class");
+		testGetPathFromURL(
+			"vfs:/opt/with%20space/tomcat/lib/servlet-api.jar/javax/servlet" +
+				"/Servlet.class",
+			"/opt/with space/tomcat/lib/servlet-api.jar/javax/servlet" +
+				"/Servlet.class");
+		testGetPathFromURL(
+			"vfs:/C:/Liferay/tomcat/lib/servlet-api.jar/javax/servlet" +
+				"/Servlet.class",
+			"/C:/Liferay/tomcat/lib/servlet-api.jar/javax/servlet" +
+				"/Servlet.class");
+		testGetPathFromURL(
+			"vfs:/C:/With%20Space/tomcat/lib/servlet-api.jar/javax/servlet" +
+				"/Servlet.class",
+			"/C:/With Space/tomcat/lib/servlet-api.jar/javax/servlet" +
+				"/Servlet.class");
+	}
+
+	@Test
+	public void testGetPathFromURLWindows() throws Exception {
+		testGetPathFromURL(
+			"jar:file:/C:/Liferay/tomcat/lib/servlet-api.jar" +
+				"!/javax/servlet/Servlet.class",
+			"/C:/Liferay/tomcat/lib/servlet-api.jar" +
```

Comment:
@jorgediaz-lr We still have the leading `/` in the expected paths. I guess it's because the unit test was written in Unix-like system which won't handle this case. If we run it in Windows, the `FileSystem` implementation will actually remove the leading `/`, and the test case will fail.

I wonder if the Windows test cases make sense to run under Linux/Mac or on CI (which only runs Linux). What do you think?

Reply:
Hi @dantewang

That code is platform independent, it will always return the path with a slash at the beginning of the path (Linux or Windows paths)
In case of Windows systems, the slash is removed after this code, using `Paths.get(URI uri)` to get a correct String path, see this code:
https://github.com/tinatian/liferay-portal/blob/526101a6e0f4c26d2be9a3156feedf6666025763/portal-kernel/src/com/liferay/portal/kernel/util/ClassUtil.java#L160-L163

If you see the final code my idea is:

1. `URL url = classLoader.getResource(className);` returns a url that is not normalized: tomcat, weblogic or wildfly return urls with different protocols or without a leading slash
2. `getPathURIFromURL(url)` normalizes the classloader URL to a normalized file URI that can be used to access to disk. This method is covered by ClassUtilTest unit tests and they can be run in both Linux and Windows without any problem
3. the URI returned in previous step, is converted to a valid String path calling `Paths.get(URI uri)`, see this commit https://github.com/tinatian/liferay-portal/commit/f5c218d14cf3f5a3326617630d759a10ec154c3b => this step will depend in the FileSystem implementation, so in case we want to add an additional unit test to this code, it won't work in Windows.

This is the only way of having:

- Unit tests that runs both Linux and Windows
- Remove the slash using Java API, without manually checking chars in positions 1 and 2 (my first proposal in previous PR)